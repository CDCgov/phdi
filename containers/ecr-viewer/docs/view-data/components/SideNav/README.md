[**ecr-viewer**](../../../README.md) â€¢ **Docs**

***

[ecr-viewer](../../../README.md) / view-data/components/SideNav

# view-data/components/SideNav

## Index

### Classes

- [SectionConfig](classes/SectionConfig.md)

### Functions

- [countObjects](functions/countObjects.md)
- [default](functions/default.md)
- [sortHeadings](functions/sortHeadings.md)
