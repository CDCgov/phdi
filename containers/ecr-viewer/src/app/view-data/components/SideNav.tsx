import React, { useState, useEffect, useRef } from "react";
import { SideNav } from "@trussworks/react-uswds";

export class SectionConfig {
  title: string;
  id: string;
  subNavItems?: SectionConfig[];

  constructor(title: string, subNavItems?: string[] | SectionConfig[]) {
    this.title = title;
    this.id = title.toLowerCase().replace(/\s+/g, "-");

    if (subNavItems) {
      this.subNavItems = subNavItems.map((item) => {
        if (typeof item === "string") {
          return new SectionConfig(item);
        } else {
          return item;
        }
      });
    }
  }
}

interface HeadingObject {
  text: string;
  level: string;
  priority: number;
}

const headingLevels = ["h1", "h2", "h3", "h4", "h5", "h6"];
const headingSelector =
  "h2:not(.unavailable-info):not(.side-nav-ignore), h3:not(.unavailable-info):not(.side-nav-ignore), h4:not(.unavailable-info):not(.side-nav-ignore)";

function countObjects(array: SectionConfig[]): number {
  let count = 0;

  function countRecursively(item: SectionConfig): void {
    count++;
    if (item.subNavItems) {
      item.subNavItems.forEach((subHead) => countRecursively(subHead));
    }
  }

  array.forEach((item) => countRecursively(item));

  return count;
}

const sortHeadings = (headings: HeadingObject[]) => {
  const result: SectionConfig[] = [];
  let i = 0;
  while (i < headings.length) {
    const currentHeading = headings[i];
    const nextHeadings = headings.slice(i + 1);
    if (
      nextHeadings.length > 0 &&
      nextHeadings[0].priority > currentHeading.priority
    ) {
      const nestedResult = sortHeadings(nextHeadings);
      result.push(new SectionConfig(currentHeading.text, nestedResult));
      const nestedLength = countObjects(nestedResult);
      i += nestedLength + 1;
    } else if (
      nextHeadings.length > 0 &&
      nextHeadings[0].priority == currentHeading.priority
    ) {
      result.push(new SectionConfig(currentHeading.text));
      i++;
    } else if (
      nextHeadings.length > 0 &&
      nextHeadings[0].priority < currentHeading.priority
    ) {
      result.push(new SectionConfig(currentHeading.text));
      i += headings.length + 1;
    } else {
      result.push(new SectionConfig(currentHeading.text));
      i++;
    }
  }
  return result;
};

const sideNav: React.FC = () => {
  const sectionRefs = useRef<{ [key: string]: HTMLDivElement | null }>({});
  const [sectionConfigs, setSectionConfigs] = useState<SectionConfig[]>([]);
  const [activeSection, setActiveSection] = useState<string>("");

  useEffect(() => {
    // Select all heading tags on the page
    const headingElements = document.querySelectorAll(headingSelector);

    // Extract the text content from each heading and store it in the state
    const headings: HeadingObject[] = Array.from(headingElements).map(
      (heading) => {
        return {
          text: heading.textContent || "",
          level: heading.tagName.toLowerCase(),
          priority: headingLevels.findIndex(
            (level) => heading.tagName.toLowerCase() == level,
          ),
        };
      },
    );
    let sortedHeadings: SectionConfig[] = sortHeadings(headings);
    setSectionConfigs(sortedHeadings);
  }, []);

  useEffect(() => {
    let options = {
      root: null,
      rootMargin: "0px 0px -80% 0px",
      threshold: 0.8,
    };

    let observer = new IntersectionObserver((entries) => {
      entries.forEach((entry) => {
        if (entry.isIntersecting) {
          let id =
            entry.target.id || entry.target.querySelectorAll("span")[0]?.id;
          console.log("id is: ", id);
          setActiveSection(id);
        }
      });
    }, options);
    const headingElements = document.querySelectorAll(headingSelector);
    headingElements.forEach((element) => observer.observe(element));
  }, []);

  function buildSideNav(sectionConfigs: SectionConfig[]) {
    let sideNavItems: React.ReactNode[] = [];
    for (let section of sectionConfigs) {
      let sideNavItem = (
        <a
          key={section.id}
          href={"#" + section.id}
          className={activeSection === section.id ? "usa-current" : ""}
        >
          {section.title}
        </a>
      );
      sideNavItems.push(sideNavItem);

      if (section.subNavItems) {
        let subSideNavItems = buildSideNav(section.subNavItems);
        sideNavItems.push(<SideNav isSubnav={true} items={subSideNavItems} />);
      }
    }

    return sideNavItems;
  }

  let sideNavItems = buildSideNav(sectionConfigs);

  return <SideNav items={sideNavItems} />;
};

export default sideNav;
