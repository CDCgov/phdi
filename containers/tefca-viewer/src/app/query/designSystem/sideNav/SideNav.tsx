import { SideNav as UswdsSideNav } from "@trussworks/react-uswds";
import { formatIdForAnchorTag } from "../../components/resultsView/ResultsViewTable";
import { ReactNode } from "react";
import styles from "./sidenav.module.css";

export type NavItem = {
  title: string;
  activeItem: boolean;
  isSubNav?: boolean;
  navItemClassName?: string;
};

export type SideNavProps = {
  items: NavItem[];
  containerClassName?: string;
  sticky?: boolean;
};

/**
 *
 * @param root0 - params
 * @param root0.items - an array of NavItems to display in the sidenav
 * @param root0.containerClassName - an optional CSS class for the sidenav container
 * @param root0.sticky - config for whether the nav is sticky
 * @returns A sidenav component
 */
const SideNav: React.FC<SideNavProps> = ({
  items,
  containerClassName,
  sticky = false,
}) => {
  const sideNavItems: ReactNode[] = [];
  for (let item of items) {
    const sideNavItem = buildSectionMarkUp(
      item.title,
      item.activeItem,
      item.navItemClassName
    );
    sideNavItems.push(
      <div className={styles.subItem}>
        <UswdsSideNav items={[sideNavItem]} isSubnav={item.isSubNav} />
      </div>
    );
  }
  return (
    <nav
      className={`${styles.container} ${containerClassName ?? ""} ${
        sticky ? styles.sticky : ""
      }`}
    >
      <UswdsSideNav items={sideNavItems} />
    </nav>
  );
};

function buildSectionMarkUp(
  title: string,
  activeItem: boolean,
  className?: string
): ReactNode {
  let sectionId = formatIdForAnchorTag(title);
  let sideNavItem = (
    <a
      key={sectionId}
      href={"#" + sectionId}
      className={`${styles.item} ${className} ${
        activeItem ? `usa-current` : ""
      }`}
    >
      {title}
    </a>
  );
  return sideNavItem;
}

export default SideNav;
