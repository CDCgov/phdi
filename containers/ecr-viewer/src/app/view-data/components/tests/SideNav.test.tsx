import SideNav, {
  SectionConfig,
  sortHeadings,
} from "@/app/view-data/components/SideNav";
import { render } from "@testing-library/react";
import { axe } from "jest-axe";

describe("SectionConfig", () => {
  beforeEach(() => {
    // IntersectionObserver isn't available in test environment
    const mockIntersectionObserver = jest.fn();
    mockIntersectionObserver.mockReturnValue({
      observe: () => null,
      unobserve: () => null,
      disconnect: () => null,
    });
    window.IntersectionObserver = mockIntersectionObserver;
  });

  it("should create an instance with correct title and id", () => {
    const section = new SectionConfig("Test Section");
    expect(section.title).toBe("Test Section");
    expect(section.id).toBe("test-section");
  });

  it("should handle subNavItems as strings and convert them to SectionConfig instances", () => {
    const section = new SectionConfig("Parent Section", ["Child Section"]);
    expect(section.subNavItems?.length).toBe(1);
    expect(section.subNavItems?.[0] instanceof SectionConfig).toBeTruthy();
    expect(section.subNavItems?.[0]?.title).toBe("Child Section");
  });

  it("should handle subNavItems as SectionConfig instances", () => {
    const childSection = new SectionConfig("Child Section");
    const section = new SectionConfig("Parent Section", [childSection]);
    expect(section.subNavItems?.length).toBe(1);
    expect(section.subNavItems?.[0]).toBe(childSection);
  });

  it("should match the snapshot", () => {
    const { asFragment } = render(
      <>
        <SideNav />
        <h2 id="section-1">Section 1</h2>
        <h2 id="section-2">Section 2</h2>
        <h3 id="section-3">Section 3</h3>
        <h4 id="section-4">Section 4</h4>
        <h2 id="section-2-2">Section 2 - 2</h2>
      </>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should have no accessibility violations", async () => {
    const { container } = render(<SideNav />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it("should sort section headings", async () => {
    const headings = [
      {
        text: "foo",
        level: "h1",
        priority: 1,
      },
      {
        text: "bar",
        level: "h2",
        priority: 2,
      },
      {
        text: "biz",
        level: "h1",
        priority: 1,
      },
    ];
    const foo = new SectionConfig("foo", ["bar"]);
    const bar = new SectionConfig("biz");

    const result: SectionConfig[] = [foo, bar];
    const resultSub = result[0]?.subNavItems;
    const sortedResults = sortHeadings(headings);
    const sortedResultsSub = sortedResults[0]?.subNavItems;
    expect(sortedResults[0].id).toBe(result[0].id);
    expect(sortedResults[1].id).toBe(result[1].id);
    expect(sortedResultsSub ? sortedResultsSub[0].id : null).toBe(
      resultSub ? resultSub[0].id : undefined,
    );
  });

  it("should only render side nav items on page", async () => {
    const { container } = render(
      <>
        <SideNav />
        <h2 id="section-1">Section 1</h2>
        <h2 id="section-2">Section 2</h2>
        <h3 id="section-3">Section 3</h3>
        <h4 id="section-4">Section 4</h4>
        <h2 id="section-2-2">Section 2 - 2</h2>
      </>,
    );
    expect(container.innerHTML).toContain('<a href="#section-1" class="">');
  });
});
