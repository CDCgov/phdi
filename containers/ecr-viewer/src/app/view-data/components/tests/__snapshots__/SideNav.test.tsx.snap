// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionConfig should match the snapshot 1`] = `
<DocumentFragment>
  <ul
    class="usa-sidenav"
    data-testid="sidenav"
  >
    <li
      class="usa-sidenav__item"
    >
      <a
        class=""
        href="#section-1"
      >
        Section 1
      </a>
    </li>
    <li
      class="usa-sidenav__item"
    >
      <a
        class=""
        href="#section-2"
      >
        Section 2
      </a>
    </li>
    <li
      class="usa-sidenav__item"
    >
      <ul
        class="usa-sidenav__sublist"
        data-testid="sidenav"
      >
        <li
          class="usa-sidenav__item"
        >
          <a
            class=""
            href="#section-3"
          >
            Section 3
          </a>
        </li>
        <li
          class="usa-sidenav__item"
        >
          <ul
            class="usa-sidenav__sublist"
            data-testid="sidenav"
          >
            <li
              class="usa-sidenav__item"
            >
              <a
                class=""
                href="#section-4"
              >
                Section 4
              </a>
            </li>
          </ul>
        </li>
        <li
          class="usa-sidenav__item"
        >
          <a
            class=""
            href="#section-2---2"
          >
            Section 2 - 2
          </a>
        </li>
      </ul>
    </li>
  </ul>
  <h2
    id="section-1"
  >
    Section 1
  </h2>
  <h2
    id="section-2"
  >
    Section 2
  </h2>
  <h3
    id="section-3"
  >
    Section 3
  </h3>
  <h4
    id="section-4"
  >
    Section 4
  </h4>
  <h2
    id="section-2-2"
  >
    Section 2 - 2
  </h2>
</DocumentFragment>
`;
